WC2-172 deduplication front end(part1)

quang-le opened the pull request:
Create views for Duplicates list and duplicates merging

Related JIRA tickets : WC2-172

## Self proofreading checklist

- [X] Did I use eslint and black formatters
- [X] Is my code clear enough and well documented - well, I think so
- [X] Are my typescript files well typed
- [X] New translations have been added or updated if new string have been introduced in the frontend
- [X] My migrations file are included
- [ ] Are there enough tests
- [ ] Documentation has been included (for new feature) -> No but I should defo write some --> I'll make a ticket

## Changes

- Add mock data because no backend yet. The shape of the payload matches what has been discussed in [WC2-171](https://bluesquare.atlassian.net/browse/WC2-171)
- Add utility hooks: `useObjectState` and `useArrayState` + unit tests for the underlying reducer functions
- Refactor `StarsComponent`: convert to TS and allow to pass more params
- Add 2 new permissions `iaso_entity_duplicates_read` and `iaso_entity_duplicates_write`
- Improve `OrgUnitTreeviewModal` to better handle `multi` mode. Add a dedicated API hook for `multi` mode
- Add `useTreeviewFilterState` hook to use with `useFilterState` when we have a treeview filter field
- Duplicates list:
    - Filters: 
        - Add all filters in mock up
        - Add filter on forms and fields. fileds's filter requires a form to be selected to populate its dropdown
        - Filter params can be seen in url when clicking the 'Filter' button
    - Table: 
        - Displays similarity score as stars
        - Highlight fields identified as mismatching by algorithm runs (colored in red)
        -  Action button redirects to details view 
 - Duplicates details
        - Has an `Info` table with some key info. "Remaining unmatched" value updates with the table
        - Merge and Ignore button redirect to list view onSuccess
        - Merge button is disabled as long as there are unmatched remaining
        - Users can hide matching fields using a check box
        - Users can auto-match with the values of either entity, or reset the values with the table buttons  
        - Selected values will appear in green, dropped values in red, unmatched in yellow.
        - You can see the query in the console when clicking 'Ignore'  or 'Merge'
        - The buttons (pre-select, ignore and merge) are only available to users with write permission

## How to test

- Go to beneficiaries -> duplicates
- Play around with the features. There's already mock data

## Print screen / video

List:

https://user-images.githubusercontent.com/38907762/225053082-85f9acf3-71e2-4892-af64-98bc16e697b9.mov


Details:

https://user-images.githubusercontent.com/38907762/225053167-98dee538-039a-41b7-873e-f41e5825cb16.mov

Permissions:


https://user-images.githubusercontent.com/38907762/225053834-e94c40d9-637b-4dc4-90a1-9c1603c1073b.mov


## Notes

⚠️Depends on https://github.com/BLSQ/bluesquare-components/pull/109⚠️
`package-lock` should be updated when the above PR is merged


[WC2-171]: https://bluesquare.atlassian.net/browse/WC2-171?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

quang-le committed:
WC2-172 WIP boilerplate for menu entry

quang-le committed:
WC2-172: add translations, path, keys

quang-le committed:
WC2-172: add pagination params to route

quang-le committed:
WC-172 add translations

quang-le committed:
WC2-172: moack GET API

quang-le committed:
WC2-172: add custom Stars component

quang-le committed:
WC2-172: add Duplicates Table

quang-le committed:
WC2-172:include all param for duplicates in routes

quang-le committed:
WC2-172: update translations

quang-le committed:
WC2-172: add DuplicatesFilters

quang-le committed:
WC2-172: update router,types, mock data for API

quang-le committed:
WC2-172: add dropdown to select inputs

quang-le committed:
WC2-172: update stars component

quang-le committed:
WC2-172: format fields label

quang-le committed:
WC2-172: add translations

quang-le committed:
WC2-172: update mocjk data and APi calls

quang-le committed:
WC2-172: add useArrayState, useObjectState utils

quang-le committed:
WC2-172: add DuplicateDetails

quang-le committed:
WC2-172: fix form column

quang-le committed:
WC2-172: demo adjustments

quang-le committed:
WC2-172: WIP add buttons to detail view

quang-le committed:
WC2-172: add DuplicatesInfos

quang-le committed:
WC2-172: add translation for TopBar

quang-le committed:
WC2-172: multiselect in org unit filter

quang-le committed:
WC2-172: refactor StarsComponent

quang-le committed:
WC2-172: export Stars style

quang-le committed:
WC2-172: enable fltering OU by status in TV hook

quang-le committed:
WC2-172: extract filter logic

quang-le committed:
Merge remote-tracking branch 'origin' into WC2-172_deduplication_front_end

quang-le committed:
Merge remote-tracking branch 'origin' into WC2-172_deduplication_front_end

quang-le committed:
WC2-172: move Details Buttons in own component

quang-le committed:
WC2-172: add translations

quang-le committed:
WC2-172: add TS typing

quang-le committed:
WC2-172: add PATCh hook (mock data)

quang-le committed:
WC2-172: add useIgnoreDuplicate

quang-le committed:
WC2-172: WIP add button to succesSnackBar

quang-le committed:
WC2-172: add Immer dep

quang-le committed:
WC2-172: update useObjectState, useArrayState

quang-le committed:
WC2-172 update translations

quang-le committed:
WC-172: update useObjectState and useArrayState

quang-le committed:
WC2-172: add typing to useGetDuplicates

quang-le committed:
WC2-172: column stayling for FF, toggle unmatched

quang-le committed:
WC2-172: depend on blsq-comp WC2-183 branch

quang-le committed:
WC2-172: add selection buttons

quang-le committed:
WC2-172: remove Immer

quang-le committed:
WC2-172: clean up typing

quang-le committed:
WC2-172: redirect onSuccess

quang-le committed:
Wc2-172: add param to disable merge button

quang-le committed:
WC2-172: add specific perms for duplicates

quang-le committed:
WC2-172: code cleanup

quang-le committed:
WC2-172: reorder files

quang-le committed:
WC2-172: code cleanup

quang-le committed:
WC2-172:add tests for useArrayState,useObjectState

quang-le committed:
Merge branch 'main' into WC2-172_deduplication_front_end

quang-le committed:
WC2-172: fix broken import path

beygorghor commented on hat/assets/js/apps/Iaso/components/stars/StarsComponent.tsx:
```diff
@@ -0,0 +1,102 @@
+import Star from './StarSvgComponent';
+
+// @ts-ignore
+const useStyles = makeStyles(theme => ({
```
`Theme` Types are available with Material ui

beygorghor commented on hat/assets/js/apps/Iaso/components/stars/StarsComponent.tsx:
```diff
@@ -0,0 +1,102 @@
+        maxScore,
+        starCount,
+    });
+    // const fullStarsRatio = (100 * fullStars) / starCount;
```
to remove

beygorghor commented on hat/assets/js/apps/Iaso/components/stars/StarsComponent.tsx:
❤️

beygorghor commented on hat/assets/js/apps/Iaso/domains/entities/duplicates/details/DuplicateDetails.tsx:
```diff
@@ -0,0 +1,284 @@
+import { DuplicateData } from '../types';
+import { useGoBack } from '../../../../routing/useGoBack';
+import { Router } from '../../../../types/general';
+import { DuplicateDetailsTableButtons } from './DuplicateDetailsTableButtons';
```
Just an idea here, on new component like this i'm trying to group import by types (component, hooks, type), It helps a lot while coming back on the file

beygorghor commented on hat/assets/js/apps/Iaso/domains/entities/duplicates/details/DuplicateDetails.tsx:
```diff
@@ -0,0 +1,284 @@
+        ...commonStyles(theme),
+        table: {
+            '& .MuiTable-root': {
+                borderLeft: `1px solid rgb(224, 224, 224)`,
+                borderRight: `1px solid rgb(224, 224, 224)`,
+                borderBottom: `1px solid rgb(224, 224, 224)`,
```
Don't we have a border color somewhere ?

quang-le replied:
> Great jobs, but a few remarks:
> 
>     * Please add a `cursor:pointer` on a clickable cell
>       ![Screenshot 2023-03-16 at 11 07 56](https://user-images.githubusercontent.com/12494624/225633891-9a9cea7e-f0cb-4020-b8ec-1aaa9ebaab4a.png)
> 
>     * it looks like a border is missing, also i would move the buttons under the infos, or outside the paper.
>       ![Screenshot 2023-03-16 at 11 07 41](https://user-images.githubusercontent.com/12494624/225634369-5eaf5daf-b809-44d4-a1f1-4f5275aa1343.png)
> 
>     * not sure for this icon
>       ![Screenshot 2023-03-16 at 11 07 29](https://user-images.githubusercontent.com/12494624/225634700-850322f0-6de6-4d87-9451-643b60402923.png)
> 
>     * The filters on the list page can take all the width, the buttons can go under
>       ![Screenshot 2023-03-16 at 11 07 14](https://user-images.githubusercontent.com/12494624/225635017-78db6e30-1e7d-4b50-a836-d8e60c6845a3.png)

- I wouldn't put the buttons under the `Paper` as it would leave tons of blank space. Also it wouldn't be consistent with other pages where such buttons are always on the right side of the screen
- About the filters: I thought the guideline was to follow that pattern:
![Screenshot 2023-03-17 at 13 44 22](https://user-images.githubusercontent.com/38907762/225908114-406de739-c4db-47f0-a828-987e28925a4c.png)
- The icon is the one used for other "compare" features, eg in instances

quang-le replied:
put it in this ticket: https://bluesquare.atlassian.net/browse/IA-1986

beygorghor commented on hat/assets/js/apps/Iaso/domains/entities/duplicates/details/DuplicateDetails.tsx:
```diff
@@ -0,0 +1,284 @@
+                    classes.containerFullHeightNoTabPadded,
+                )}
+            >
+                <Grid container>
+                    <Grid item xs={12}>
```
Grid looks useless here

beygorghor commented on hat/assets/js/apps/Iaso/domains/entities/duplicates/details/DuplicateDetailsTableButtons.tsx:
```diff
@@ -0,0 +1,84 @@
+                <Grid container item xs={8} justifyContent="flex-end">
+                    <Box
+                        pb={2}
+                        pt={2}
```
you can use `py` prop here see https://v4.mui.com/system/spacing/#api

beygorghor commented on hat/assets/js/apps/Iaso/domains/entities/duplicates/details/DuplicateInfos.tsx:
```diff
@@ -0,0 +1,142 @@
+    const classes: Record<string, string> = useStyles();
+    const dispatch = useDispatch();
+    const successSnackBar = (msg, data) => {
+        return succesfullSnackBar(formSuccessFullMessageKey, msg);
```
We can have a look together on this

quang-le replied:
Yeah, it would be a cool feature to have

quang-le replied:
Jira: https://bluesquare.atlassian.net/browse/IA-1987

beygorghor commented on hat/assets/js/apps/Iaso/domains/entities/duplicates/details/DuplicateInfosTable.tsx:
```diff
@@ -0,0 +1,66 @@
+}) => {
+    const { formatMessage } = useSafeIntl();
+    return (
+        <Table>
```
Maybe use the small version of the table, like we did on other pages

beygorghor commented on hat/assets/js/apps/Iaso/domains/entities/duplicates/details/hooks/useDuplicationDetailsColumns.tsx:
```diff
@@ -0,0 +1,106 @@
+
+                    return (
+                        // eslint-disable-next-line jsx-a11y/no-static-element-interactions
+                        <div onClick={onClick}>
```
Just add role and tab index on static element to make it clickable

beygorghor commented on hat/assets/js/apps/Iaso/domains/entities/duplicates/hooks/api/useGetDuplicates.ts:
```diff
@@ -0,0 +1,150 @@
+
+const getDuplicates = async (queryString: string) => {
+    const url = `${apiUrl}/${queryString}`;
+    console.log('GET', url);
```
to remove

quang-le replied:
You mean the console.log? Actually it should stay until we have the backend. That way we can show that the query is updated according to filter values.

beygorghor commented on hat/assets/js/apps/Iaso/domains/entities/duplicates/list/DuplicatesFilters.tsx:
```diff
@@ -0,0 +1,254 @@
+    }, [filters.fields, filters.form, handleChange]);
+
+    return (
+        <>
```
fragment not necessary

beygorghor commented on hat/assets/js/apps/Iaso/domains/entities/duplicates/list/useDuplicationTableColumns.tsx:
```diff
@@ -0,0 +1,158 @@
+                            })}
+                        </>
+                    );
+                },
```
Those two ccell can be merged into one component no ? entityA and entityB

beygorghor commented on hat/assets/js/apps/Iaso/hooks/useArrayState.spec.js:
❤️

beygorghor replied:
didi you check the coverage of this ? `npm run test-coverage` it will generate a report in html

beygorghor reviewed (changes requested):
Great jobs, but a few remarks:

- Please add a `cursor:pointer` on a clickable cell
![Screenshot 2023-03-16 at 11 07 56](https://user-images.githubusercontent.com/12494624/225633891-9a9cea7e-f0cb-4020-b8ec-1aaa9ebaab4a.png)
- it looks like a border is missing, also i would move the buttons under the infos, or outside the paper.
![Screenshot 2023-03-16 at 11 07 41](https://user-images.githubusercontent.com/12494624/225634369-5eaf5daf-b809-44d4-a1f1-4f5275aa1343.png)
- not sure for this icon
![Screenshot 2023-03-16 at 11 07 29](https://user-images.githubusercontent.com/12494624/225634700-850322f0-6de6-4d87-9451-643b60402923.png)
- The filters on the list page can take all the width, the buttons can go under
![Screenshot 2023-03-16 at 11 07 14](https://user-images.githubusercontent.com/12494624/225635017-78db6e30-1e7d-4b50-a836-d8e60c6845a3.png)

quang-le committed:
WC2-172: code review (1)

quang-le committed:
WC2-172: improve Details page (code review)

- code review improvemeents
- fix bug with show Only Unmatched and selection buttons

quang-le committed:
WC2-172: rearrange filters
